### ISSUE TITLE
GPU memory consumption keeps increasing with multithreading in Java

### ISSUE THREAD

**Issue by andyguo007** [microsoft/onnxruntime#6181]
**Describe the bug**
Hi team,
When invoking the run method in a for loop in one thread using the Java API, the GPU memory consumption stays at a low level and unchanged, which is expected.

However, when using a thread pool with multiple threads (e.g., 30 threads) to invoke the Run() method on the same inference session object simultaneously, the GPU memory consumption keeps increasing until all the GPU memory is used up. Is this expected?

**Urgency**
If there are particular important use cases blocked by this or strict project-related timelines, please share more information and dates. If there are no hard deadlines, please specify none.
None

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- ONNX Runtime installed from (source or binary): source
- ONNX Runtime version: 1.6.0
- Python version:
- Visual Studio version (if applicable):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: CUDA 11.0, cuDNN 8.0.5
- GPU model and memory: V100, 16G

**To Reproduce**
- Describe steps/code to reproduce the behavior.
- Attach the ONNX model to the issue (where applicable) to expedite investigation.

Some code snippet below that would cause the all the GPU memory to be used:

```
Mat mat = Imgcodecs.imread(inputImagePath, Imgcodecs.CV_LOAD_IMAGE_COLOR);
int numOfThreads = 30;
int trials = 30;
Executor executor = Executors.newFixedThreadPool(numOfThreads);
CountDownLatch countDownLatch = new CountDownLatch(trials);
for (int i = 0; i < trials; i++) {
    executor.execute(() -> {
        ObjectDetection.detectObjectsYoloV4(mat);
        countDownLatch.countDown();
    });
}

try {
    countDownLatch.await();
} catch (InterruptedException e) {
    e.printStackTrace();
}
```

**Comment by Craigacp** [comment]
Hmm, that's odd as Java should only be interacting with the CPU memory because we currently can't pin things in GPU memory. The native memory inside the tensor should be freed when the `OnnxTensor.close()` method is called. 

What does the object detection call do? What types are output by the call to run on that yolo model?

**Comment by andyguo007** [comment]
> Hmm, that's odd as Java should only be interacting with the CPU memory because we currently can't pin things in GPU memory. The native memory inside the tensor should be freed when the `OnnxTensor.close()` method is called.
> 
> What does the object detection call do? What types are output by the call to run on that yolo model?

Thanks @Craigacp . The main body of the ```ObjectDetection.detectObjectsYoloV4(mat)``` is as below. I also tried YoloV5, and the same issue exists. I could share the model with you if you would like to try.

```
OrtSession.Result output = null;
try (OnnxTensor inputTensor = makeImageTensor(ortEnvironment, chwPixels, 1, channels, height, width)) {
    output = yolov4Session.run(Collections.singletonMap("input", inputTensor));
   
    float[][][] boxes = ((float[][][][])output.get(0).getValue())[0];
    float[][] scores = ((float[][][])output.get(1).getValue())[0];  
    // some other logic to use boxes and scores to construct detected objects goes here
} finally {
    if (output != null) {
        output.close();
    }
}
```

**Comment by Craigacp** [comment]
Yes, if you could send the model that would be helpful. You can have the result object live inside the try with resources as well if you want which would ensure it's closed without the finally statement, but that shouldn't cause any leaks.

**Comment by andyguo007** [comment]
> Yes, if you could send the model that would be helpful. You can have the result object live inside the try with resources as well if you want which would ensure it's closed without the finally statement, but the way it is in that snippet shouldn't cause any leaks.

The model is uploade [here](https://drive.google.com/drive/folders/1BKgSDNbKifni4Wg8ztLrhbok0nM9IFrK?usp=sharing), I also created some code snippet there for your reference. Thanks.

**Comment by andyguo007** [comment]
Any updates? @Craigacp
